src/dmd/func.d

```diff
@@ -2748,8 +2748,17 @@ extern (C++) FuncDeclaration resolveFuncCall(const ref Loc loc, Scope* sc, Dsymb
         {
             assert(fd);
 
+            uint errors;
+            bool isCpCtor = fd.isCopyCtorDeclaration() !is null;
+            if (isCpCtor)
+                errors = global.startGagging();
+
             if (fd.checkDisabled(loc, sc))
+            {
+                if (isCpCtor)
+                    global.endGagging(errors);
```

Comment:
This is awfully hacky. checkDisabled() should accept a parameter that says whether to print the error or not.

Reply:
In fact, just replace the checkDisabled() call with `fd.storage_class & STC.disable`, as the rest of the function is either irrelevant or entirely concerned with printing the message, and you don't want to print the message. Then, you can get rid of all the gagging crud.

Reply:
It's not just that. If there is no copy constructor suitable for the initialization, trivial copying is tried before issuing an error. This is done to avoid the following scenario : 

```d
struct A
{
    @implicit this(ref shared A rhs) immutable {}
} 

void main()
{
    A a;
    A b = a;     // error : no copy constructor for mutable-mutable copies
}
```

The rule of thumb is : see if there is a valid copy constructor to be called; if yes, call it; if not, try to employ trivial copying.

Reply:
I don't think that generating a trivial cpctor is the right move if there is an existing one that doesn't match. It should produce an error instead - otherwise the user may erroneously assume it is calling the cpctor he wrote.

Reply:
@WalterBright this point is well argued in the DIP. If defining one ctor disables all implicitly-defined ones, the user would need to define up to EIGHT copy constructors - a hecatomb of code. The proposed behavior is to let the user define the exceptional cases and let the default behavior take care of the simple ones.